src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -862,185 +768,335 @@ private static HttpRequestException CreateOverCapacityException(long maxBufferSi
         }
 
-        internal sealed class LimitMemoryStream : MemoryStream
+        /// <summary>
+        /// A write-only stream that limits the total length of the content to <see cref="_maxBufferSize"/>.
+        /// It uses pooled buffers for the content, but can switch to a regular array allocation, which is useful when the caller
+        /// already knows the final size and needs a new array anyway (e.g. <see cref="HttpClient.GetByteArrayAsync(string?)"/>).
+        /// <para>Since we can't rely on users to reliably dispose content objects, any pooled buffers must be returned before leaving
+        /// the execution path we control. In practice this means <see cref="LoadIntoBufferAsync()"/> must call <see cref="ReturnAllPooledBuffers"/>
+        /// before storing the stream in <see cref="_bufferedContent"/>.</para>
+        /// </summary>
+        internal sealed class LimitArrayPoolWriteStream : Stream
         {
-            private readonly int _maxSize;
+            /// <summary>Applies when a Content-Length header was not specified.
+            /// <para>These buffers are pooled and we're only holding onto them while downloading, so we start with a relatively large size.</para></summary>
+            private const int MinInitialBufferSize = 256 * 1024; // 256 KB
+
+            /// <summary>Applies when a Content-Length header was set. If it's &lt;= this limit, we'll allocate an exact-sized buffer upfront.</summary>
+            private const int MaxInitialBufferSize = 16 * 1024 * 1024; // 16 MB
+
+            /// <summary>Controls how quickly we're willing to expand up to <see cref="_expectedFinalSize"/> when a caller requested that the last buffer should not be pooled.
+            /// <para>The factor is higher than usual to lower the number of memory copies when the caller already committed to allocating a large buffer.</para></summary>
+            private const int LastResizeFactor = 4;
 
-            public LimitMemoryStream(int maxSize, int capacity)
-                : base(capacity)
+            /// <summary><see cref="_totalLength"/> may not exceed this limit, or we'll throw a <see cref="CreateOverCapacityException"/>.</summary>
+            private readonly int _maxBufferSize;
+            /// <summary>The value of the Content-Length header or 0. <see cref="_totalLength"/> may exceed this value if the Content-Length isn't being enforced by the content.</summary>
+            private readonly int _expectedFinalSize;
+            /// <summary>Indicates whether the caller will need an exactly-sized, non-pooled, buffer. The implementation will switch away from pooled buffers earlier to reduce memory copies.</summary>
+            private readonly bool _shouldPoolFinalSize;
+
+            private bool _lastBufferIsPooled;
+            private byte[] _lastBuffer;
+            private byte[]?[]? _pooledBuffers;
```

Comment:
Would it be possible to leverage [`MutliArrayBuffer`](https://github.com/dotnet/runtime/blob/d3d0fced1d125483d668bef176618e6a805cb2bd/src/libraries/Common/src/System/Net/MultiArrayBuffer.cs), which already has a collection of pooled buffers -- or will that be terribly inefficient?

Or the other way around: could something from here improve the MutliArrayBuffer implementation?

It just feels a little bit like reimplementing a very similar idea, but maybe I'm missing something.

Reply:
It's a good question and to a degree we are doing very similar things.
Using `MultiArrayBuffer` as-is would indeed be inefficient when it comes to larger responses because it uses only one fixed size for buffers (16 KB). A large content could quickly negate any pooling benefits by starving out the array pool for that bucket size.

That shouldn't be as much of an issue with our current uses of `MultiArrayBuffer` because those are generally meant to be short-lived and relatively small (e.g. hand off point between the connection's read loop and the request's calls to `ReadAsync`).
Side note: if we wanted to get really fancy, we could try to avoid that extra memory copy and instead rent new buffers for connection's reads, but that's something for the (far) future.

I think it's more likely we would go the other way as you mention and tweak the `MultiArrayBuffer` implementation instead. Its use of a single buffer size seems quite ingrained in most of the implementation, but looking more at our current usages of it, it shouldn't be necessary (for example, we don't care about faster indexing into the buffer that that allows for). I'd be more inclined to replace `MultiArrayBuffer` entirely with something closer to the buffering logic here. It might be easier to reason about the change as a follow-up PR though (at risk of having two somewhat duplicatey concepts in the repo for now).
The implementation I'm adding in `HttpContent` does have some odd specifics around its buffer growth logic, and optional switching from pooled to non-pooled buffers. That part probably isn't really transferable to other uses, but I think we can make a shared implementation work without regressing those.